File: Final/Code/Project_4/BaiToan6/Divide_and_conquer/Divide_and_conquer.cpp
#include <iostream>
#include <string>
#include <algorithm>
#include <cmath>

struct TreeNode {
    std::string label;
    TreeNode* left;
    TreeNode* right;

    TreeNode(std::string lbl, TreeNode* l = nullptr, TreeNode* r = nullptr)
        : label(lbl), left(l), right(r) {}
};

int count_nodes(TreeNode* t) {
    if (!t) return 0;
    return 1 + count_nodes(t->left) + count_nodes(t->right);
}

int tree_edit_distance_dc(TreeNode* t1, TreeNode* t2) {
    if (!t1 && !t2) return 0;
    if (!t1) return count_nodes(t2);
    if (!t2) return count_nodes(t1);

    int cost = (t1->label == t2->label) ? 0 : 1;

    int cost1 = tree_edit_distance_dc(t1->left, t2->left) +
                tree_edit_distance_dc(t1->right, t2->right);
    int cost2 = tree_edit_distance_dc(t1->left, t2->right) +
                tree_edit_distance_dc(t1->right, t2->left);

    return cost + std::min(cost1, cost2);
}

int main() {
    TreeNode* t1 = new TreeNode("A", new TreeNode("B"), new TreeNode("C"));
    TreeNode* t2 = new TreeNode("A", new TreeNode("B"), new TreeNode("D"));

    std::cout << "TED = " << tree_edit_distance_dc(t1, t2) << std::endl; 

    delete t1->left; delete t1->right; delete t1;
    delete t2->left; delete t2->right; delete t2;
}

File: Final/Code/Project_4/BaiToan6/Divide_and_conquer/Divide_and_conquer.py
class TreeNode:
    def __init__(self, label, left=None, right=None):
        self.label = label
        self.left = left
        self.right = right

def count_nodes(t):
    if t is None:
        return 0
    return 1 + count_nodes(t.left) + count_nodes(t.right)

def tree_edit_distance_dc(t1, t2):
    if t1 is None and t2 is None:
        return 0
    if t1 is None:
        return count_nodes(t2)
    if t2 is None:
        return count_nodes(t1)

    cost = 0 if t1.label == t2.label else 1

    cost1 = tree_edit_distance_dc(t1.left, t2.left) + tree_edit_distance_dc(t1.right, t2.right)
    cost2 = tree_edit_distance_dc(t1.left, t2.right) + tree_edit_distance_dc(t1.right, t2.left)

    return cost + min(cost1, cost2)
# Test
if __name__ == "__main__":
    t1 = TreeNode("A", TreeNode("B"), TreeNode("C"))

    t2 = TreeNode("A", TreeNode("B"), TreeNode("D"))

    print("TED =", tree_edit_distance_dc(t1, t2))  
